File: selenium/TC-func-REQ-SAF-003-15.java
import org.openqa.selenium.*;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.support.ui.*;
import org.testng.Assert;
import org.testng.annotations.*;
import java.time.Duration;
import java.util.concurrent.TimeUnit;

public class TC-func-REQ-SAF-003-15_Test {

    private WebDriver driver;
    private WebDriverWait wait;

    @BeforeMethod
    public void setUp() {
        driver = new ChromeDriver();
        driver.manage().timeouts().implicitlyWait(10, TimeUnit.SECONDS);
        wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        // Navigate to the application under test
        driver.get("http://your-vehicle-dashboard-url.com");
    }

    @Test
    public void test_TC-func-REQ-SAF-003-15() throws InterruptedException {
        // Step 1: Induce communication glitch
        // Simulate a 5-second communication glitch
        Thread.sleep(5000);

        // Verify low fuel warning remains active during glitch
        By lowFuelWarning = By.cssSelector("[data-testid='low-fuel-warning']");
        wait.until(ExpectedConditions.visibilityOfElementLocated(lowFuelWarning));
        Assert.assertTrue(driver.findElement(lowFuelWarning).isDisplayed(),
                "Low fuel warning should remain active during communication glitch");

        // Step 2: Restore communication
        // (No action needed in this simulation)

        // Verify low fuel warning still active after restoring communication
        wait.until(ExpectedConditions.visibilityOfElementLocated(lowFuelWarning));
        Assert.assertTrue(driver.findElement(lowFuelWarning).isDisplayed(),
                "Low fuel warning should remain active after communication is restored");

        // Step 3: Increase fuel level to 121 km
        // Locate fuel level input or button and set value to 121
        By fuelLevelInput = By.cssSelector("[data-testid='fuel-level-input']");
        WebElement fuelInput = wait.until(ExpectedConditions.elementToBeClickable(fuelLevelInput));
        fuelInput.clear();
        fuelInput.sendKeys("121");
        // Optionally submit or trigger update
        By updateButton = By.cssSelector("[data-testid='fuel-level-update']");
        WebElement updateBtn = wait.until(ExpectedConditions.elementToBeClickable(updateButton));
        updateBtn.click();

        // Verify low fuel warning clears after threshold is crossed
        wait.until(ExpectedConditions.invisibilityOfElementLocated(lowFuelWarning));
        Assert.assertFalse(driver.findElements(lowFuelWarning).size() > 0,
                "Low fuel warning should clear after fuel level exceeds threshold");
    }

    @AfterMethod
    public void tearDown() {
        if (driver != null) {
            driver.quit();
        }
    }
}

File: playwright/TC-func-REQ-SAF-003-15.ts
import { test, expect } from '@playwright/test';

test("TC-func-REQ-SAF-003-15 - Verify warning state persistence during transient communication glitch", async ({ page }) => {
  // Navigate to vehicle dashboard
  await page.goto('/vehicle-dashboard');

  // Step 1: Induce communication glitch
  await page.locator('[data-testid="simulate-comm-glitch"]').click();
  await page.waitForTimeout(5000); // simulate 5-second glitch
  const warning = page.locator('[data-testid="low-fuel-warning"]');
  await expect(warning).toBeVisible();

  // Step 2: Restore communication
  await page.locator('[data-testid="restore-comm"]').click();
  await expect(warning).toBeVisible();

  // Step 3: Increase fuel level to 121 km
  await page.locator('[data-testid="fuel-level-input"]').fill('121');
  await page.locator('[data-testid="update-fuel"]').click();
  await expect(warning).not.toBeVisible();
});